// @envoy/envoy-integrations-sdk/dist/factories/entryEventBodyFactory.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.entryPayloadFactory = exports.entryEventBodyFactoryDefaultIds = void 0;
var factory_ts_1 = require("factory.ts");
var faker_1 = __importDefault(require("faker"));
var eventBodyFactory_1 = __importDefault(require("./eventBodyFactory"));
exports.entryEventBodyFactoryDefaultIds = {
    location: '1',
    company: '1',
    flow: '1',
    invite: '1',
    device: '1',
    employee: '1',
};
function entryPayloadFactory(options, ids) {
    if (ids === void 0) { ids = exports.entryEventBodyFactoryDefaultIds; }
    var signedInDate = faker_1.default.date.past();
    var allIds = __assign(__assign({}, exports.entryEventBodyFactoryDefaultIds), ids);
    return factory_ts_1.Sync.makeFactory({
        id: factory_ts_1.each(function (i) { return "" + (i + 1); }),
        type: 'entries',
        attributes: {
            'full-name': faker_1.default.name.findName(),
            'phone-number': options.hasPhoneNumber ? faker_1.default.phone.phoneNumber() : undefined,
            email: (options.isProtectFlow || options.hasEmail) ? faker_1.default.internet.email() : null,
            'employee-screening-flow': options.isProtectFlow,
            host: (!options.isProtectFlow && options.hasHost) ? faker_1.default.name.findName() : null,
            'host-email': (!options.isProtectFlow && options.hasHost) ? faker_1.default.internet.email() : null,
            'private-notes': null,
            'signed-in-at': signedInDate.toISOString(),
            'signed-out-at': options.isSignedIn ? undefined : faker_1.default.date.between(signedInDate, new Date()).toISOString(),
            thumbnails: options.hasPhoto ? {
                large: faker_1.default.image.avatar(),
                original: faker_1.default.image.avatar(),
                small: faker_1.default.image.avatar(),
            } : {
                large: null,
                original: null,
                small: null,
            },
            'flow-name': options.isProtectFlow ? 'Employee registration' : 'Visitor',
            'user-data': [],
        },
        relationships: {
            location: {
                data: {
                    id: allIds.location,
                    type: 'locations',
                },
            },
            flow: {
                data: {
                    id: allIds.flow,
                    type: 'flows',
                },
            },
            invite: (options.isProtectFlow || options.hasInvite) ? {
                data: {
                    id: allIds.invite,
                    type: 'invites',
                },
            } : undefined,
            device: (!options.isProtectFlow && options.hasDevice) ? {
                data: {
                    id: allIds.device,
                    type: 'devices',
                },
            } : undefined,
            employee: (options.isProtectFlow || options.hasHost) ? {
                data: {
                    id: allIds.employee,
                    type: 'employees',
                },
            } : undefined,
        },
    });
}
exports.entryPayloadFactory = entryPayloadFactory;
function entryEventBodyFactory(options) {
    var _a, _b;
    var ids = options.ids || exports.entryEventBodyFactoryDefaultIds;
    return eventBodyFactory_1.default({
        event: options.event,
        config: options.config,
        scope: options.scope || [],
        locationId: ((_a = options.ids) === null || _a === void 0 ? void 0 : _a.location) || exports.entryEventBodyFactoryDefaultIds.location,
        companyId: ((_b = options.ids) === null || _b === void 0 ? void 0 : _b.company) || exports.entryEventBodyFactoryDefaultIds.company,
        payload: entryPayloadFactory(options.payloadOptions, ids).build(),
    });
}
exports.default = entryEventBodyFactory;
