// @envoy/envoy-integrations-sdk/dist/base/EnvoyPluginStoragePipeline.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Builds up a request pipeline for submitting storage commands.
 *
 * @category Base
 */
var EnvoyPluginStoragePipeline = /** @class */ (function () {
    function EnvoyPluginStoragePipeline(pluginAPI, installId) {
        this.api = pluginAPI;
        this.installId = installId;
        this.commands = [];
    }
    /**
     * Executes all the commands in the pipeline.
     */
    EnvoyPluginStoragePipeline.prototype.execute = function () {
        return this.api.storagePipeline(this.commands, this.installId);
    };
    /**
     * Executes the pipeline and returns the first result.
     */
    EnvoyPluginStoragePipeline.prototype.executeSingle = function () {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.execute()];
                    case 1:
                        result = (_a.sent())[0];
                        return [2 /*return*/, result];
                }
            });
        });
    };
    EnvoyPluginStoragePipeline.prototype.addCommand = function (command) {
        this.commands.push(command);
        return this;
    };
    /**
     * Gets a storage item.
     */
    EnvoyPluginStoragePipeline.prototype.get = function (key) {
        return this.addCommand({ action: 'get', key: key });
    };
    /**
     * Sets a value for a storage item,
     * and returns that item.
     */
    EnvoyPluginStoragePipeline.prototype.set = function (key, value) {
        return this.addCommand({ action: 'set', key: key, value: value });
    };
    /**
     * Sets a unique value for a storage item,
     * and returns that item.
     */
    EnvoyPluginStoragePipeline.prototype.setUnique = function (key, options) {
        if (options === void 0) { options = {}; }
        return this.addCommand(__assign({ action: 'set_unique', key: key }, options));
    };
    /**
     * Sets a unique number value for a storage item,
     * and returns that item.
     */
    EnvoyPluginStoragePipeline.prototype.setUniqueNum = function (key, options) {
        if (options === void 0) { options = {}; }
        if (options.min) {
            options.min = Number(options.min);
        }
        if (options.max) {
            options.max = Number(options.max);
        }
        return this.addCommand(__assign({ action: 'set_unique_num', key: key }, options));
    };
    /**
     * Unsets a storage item.
     */
    EnvoyPluginStoragePipeline.prototype.unset = function (key) {
        return this.addCommand({ action: 'unset', key: key });
    };
    /**
     * Lists storage items.
     */
    EnvoyPluginStoragePipeline.prototype.list = function (page) {
        if (page === void 0) { page = 1; }
        if (!page) {
            page = 1;
        }
        return this.addCommand({ action: 'list', page: page });
    };
    return EnvoyPluginStoragePipeline;
}());
exports.default = EnvoyPluginStoragePipeline;
