// @envoy/envoy-integrations-sdk/dist/util/axiosConstructor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAxiosClient = exports.sanitizeAxiosError = void 0;
var axios_1 = __importStar(require("axios"));
var errorHandling_1 = require("./errorHandling");
function sanitizeAxiosError(error) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
    if (!axios_1.default.isAxiosError(error)) {
        return errorHandling_1.ensureError(error);
    }
    var safeError = new axios_1.AxiosError(error.message, error.code, {
        url: (_a = error.config) === null || _a === void 0 ? void 0 : _a.url,
        method: (_b = error.config) === null || _b === void 0 ? void 0 : _b.method,
        baseURL: (_c = error.config) === null || _c === void 0 ? void 0 : _c.baseURL,
        headers: new axios_1.AxiosHeaders(),
    }, {
        url: (_d = error.request) === null || _d === void 0 ? void 0 : _d.url,
        method: (_e = error.request) === null || _e === void 0 ? void 0 : _e.method,
        baseURL: (_f = error.request) === null || _f === void 0 ? void 0 : _f.baseURL,
        data: (_g = error.request) === null || _g === void 0 ? void 0 : _g.data,
    }, {
        data: (_h = error.response) === null || _h === void 0 ? void 0 : _h.data,
        status: (_k = (_j = error.response) === null || _j === void 0 ? void 0 : _j.status) !== null && _k !== void 0 ? _k : 0,
        statusText: (_m = (_l = error.response) === null || _l === void 0 ? void 0 : _l.statusText) !== null && _m !== void 0 ? _m : '',
        headers: new axios_1.AxiosHeaders(),
        config: {
            headers: new axios_1.AxiosHeaders(),
        },
    });
    safeError.stack = error.stack;
    return safeError;
}
exports.sanitizeAxiosError = sanitizeAxiosError;
function createAxiosClient(config) {
    var client = axios_1.default.create(config);
    client.interceptors.response.use(function (response) { return response; }, function (error) { return Promise.reject(sanitizeAxiosError(error)); });
    return client;
}
exports.createAxiosClient = createAxiosClient;
